src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs

```diff
@@ -19,255 +18,99 @@ namespace Microsoft.CodeAnalysis.PatternMatching
+            CultureInfo culture = null,
+            bool includeMatchedSpans = false,
+            bool allowFuzzyMatching = false)
         {
-            return _invalidPattern || string.IsNullOrWhiteSpace(candidate);
+            return new SimplePatternMatcher(pattern, culture, includeMatchedSpans, allowFuzzyMatching);
         }
 
-        public ImmutableArray<PatternMatch> GetMatches(string candidate)
-            => GetMatches(candidate, includeMatchSpans: false);
-
-        /// <summary>
-        /// Determines if a given candidate string matches under a multiple word query text, as you
-        /// would find in features like Navigate To.
-        /// </summary>
-        /// <param name="candidate">The word being tested.</param>
-        /// <param name="includeMatchSpans">Whether or not the matched spans should be included with results</param>
-        /// <returns>If this was a match, a set of match types that occurred while matching the
-        /// patterns. If it was not a match, it returns null.</returns>
-        public ImmutableArray<PatternMatch> GetMatches(string candidate, bool includeMatchSpans)
+        public static PatternMatcher CreateContainerPatternMatcher(
+            string[] patternParts,
+            char[] containerSplitCharacters,
+            CultureInfo culture = null,
+            bool allowFuzzyMatching = false)
         {
-            if (SkipMatch(candidate))
-            {
-                return ImmutableArray<PatternMatch>.Empty;
-            }
-
-            var result = MatchPatternSegment(candidate, includeMatchSpans, _fullPatternSegment, fuzzyMatch: true);
-            if (!result.IsEmpty)
-            {
-                return result;
-            }
-
-            return MatchPatternSegment(candidate, includeMatchSpans, _fullPatternSegment, fuzzyMatch: false);
+            return new ContainerPatternMatcher(
+                patternParts, containerSplitCharacters, culture, allowFuzzyMatching);
         }
 
-        public ImmutableArray<PatternMatch> GetMatchesForLastSegmentOfPattern(string candidate)
+        public static PatternMatcher CreateDotSeperatedContainerMatcher(
+            string pattern,
+            CultureInfo culture = null,
+            bool allowFuzzyMatching = false)
         {
-            if (SkipMatch(candidate))
-            {
-                return ImmutableArray<PatternMatch>.Empty;
-            }
-
-            var result = MatchPatternSegment(candidate, includeMatchSpans: false, patternSegment: _dotSeparatedPatternSegments.Last(), fuzzyMatch: false);
-            if (!result.IsEmpty)
-            {
-                return result;
-            }
-
-            return MatchPatternSegment(candidate, includeMatchSpans: false, patternSegment: _dotSeparatedPatternSegments.Last(), fuzzyMatch: true);
+            return CreateContainerPatternMatcher(
+                pattern.Split(s_dotCharacterArray), s_dotCharacterArray, culture, allowFuzzyMatching);
```

Comment:
I couldn't think of a case where it mattered.  But i've added it back in in case there's somethign i missed.